cranelift/assembler-x64/src/api.rs

```diff
@@ -0,0 +1,156 @@
+
+    /// Return the byte offset of the current location in the code buffer;
+    /// required for assembling RIP-relative memory accesses.
+    fn current_offset(&self) -> u32;
+
+    /// Inform the code buffer of a use of `label` at `offset`; required for
+    /// assembling RIP-relative memory accesses.
+    fn use_label_at_offset(&mut self, offset: u32, label: Label);
+
+    /// Return the label for a constant `id`; required for assembling
+    /// RIP-relative memory accesses of constants.
+    fn get_label_for_constant(&mut self, id: Constant) -> Label;
+}
+
+/// Provide a convenient implementation for testing.
+impl CodeSink for Vec<u8> {
+    fn put1(&mut self, v: u8) {
+        self.extend_from_slice(&[v]);
+    }
+
+    fn put2(&mut self, v: u16) {
+        self.extend_from_slice(&v.to_le_bytes());
+    }
+
+    fn put4(&mut self, v: u32) {
+        self.extend_from_slice(&v.to_le_bytes());
+    }
+
+    fn put8(&mut self, v: u64) {
+        self.extend_from_slice(&v.to_le_bytes());
+    }
+
+    fn add_trap(&mut self, _: TrapCode) {}
+
+    fn current_offset(&self) -> u32 {
+        self.len().try_into().unwrap()
+    }
+
+    fn use_label_at_offset(&mut self, _: u32, _: Label) {}
+
+    fn get_label_for_constant(&mut self, c: Constant) -> Label {
+        Label(c.0)
+    }
+}
+
+/// Wrap [`CodeSink`]-specific labels.
+#[derive(Debug, Clone, Arbitrary)]
+pub struct Label(pub u32);
+
+/// Wrap [`CodeSink`]-specific constant keys.
+#[derive(Debug, Clone, Arbitrary)]
+pub struct Constant(pub u32);
+
+/// Wrap [`CodeSink`]-specific trap codes.
+#[derive(Debug, Clone, Copy, Arbitrary)]
+pub struct TrapCode(pub NonZeroU8);
+
+/// A table mapping `KnownOffset` identifiers to their `i32` offset values.
+///
+/// When encoding instructions, Cranelift may not know all of the information
+/// needed to construct an immediate. Specifically, addressing modes that
+/// require knowing the size of the tail arguments or outgoing arguments (see
+/// `SyntheticAmode::finalize`) will not know these sizes until emission.
+///
+/// This table allows up to do a "late" look up of these values by their
+/// `KnownOffset`.
+pub trait KnownOffsetTable: Index<KnownOffset, Output = i32> {}
+impl KnownOffsetTable for Vec<i32> {}
+/// Provide a convenient implementation for testing.
+impl KnownOffsetTable for [i32; 2] {}
+
+/// A `KnownOffset` is a unique identifier for a specific offset known only at
+/// emission time.
+pub type KnownOffset = usize;
+
+/// A type set fixing the register types used in the assembler.
+///
+/// This assembler is parameterizable over register types; this allows the
+/// assembler users (e.g., Cranelift) to define their own register types
+/// independent of this crate.
+pub trait Registers {
+    /// An x64 general purpose register that may be read.
+    type ReadGpr: AsReg + for<'a> Arbitrary<'a>;
+
+    /// An x64 general purpose register that may be read and written.
+    type ReadWriteGpr: AsReg + for<'a> Arbitrary<'a>;
```

Comment:
If possible could the `Arbitrary` methods be left off? And the `Arbitrary` impl for each instruction have `where R::ReadGpr: Arbitrary`?

Reply:
(I pushed up a commit for this over at https://github.com/abrown/wasmtime/pull/10)

Reply:
Thanks!